feat: resolve endpoint target CNAME to A/AAAA records with external-dns.alpha.kubernetes.io/resolve-target (#6329)
kubernetes-sigs/external-dns: docs/annotations/annotations.md

The following table documents which sources support which annotations:

> | Source       | controller | hostname | internal-hostname | resolve-target | target  | ttl     | (provider-specific) |

Author: removed in 945ae069d4c71d6245de460ad315fa476c7c6a0c, added to docs in 01a84ef572388fb742da6756cd9518d777e6e46d

| Kong         |            | Yes[^1]  |                   |                | Yes     | Yes     | Yes                 |
| Node         | Yes        |          |                   |                | Yes     | Yes     |                     |
| OpenShift    | Yes        | Yes[^1]  |                   |                | Yes     | Yes     | Yes                 |
| Pod          |            | Yes      | Yes               |                | Yes     |         |                     |
| Service      | Yes        | Yes[^1]  | Yes[^1][^2]       |                | Yes[^3] | Yes     | Yes                 |
| Skipper      | Yes        | Yes[^1]  |                   |                | Yes     | Yes     | Yes                 |
| Traefik      |            | Yes[^1]  |                   |                | Yes     | Yes     | Yes                 |

[^1]: Unless the `--ignore-hostname-annotation` flag is specified.
[^2]: Only behaves differently than `hostname` for `Service`s of type `ClusterIP` or `LoadBalancer`.
[^3]: Also supported on `Pods` referenced from a headless `Service`'s `Endpoints`.

> [^4]: For Gateway API sources, annotation placement differs by type. See [Gateway API Annotation Placement](#gateway-api-annotation-placement) for details. `resolve-target` must be placed on the **Route** resource.

Author: Yes, the annotation is required on the route resource, as we don't have a "gateway" source. This is added in https://github.com/kubernetes-sigs/external-dns/pull/6380. If that PR gets merged it will also work on the Gateway resource and will create a dns for a "Gateway"

> ExternalDNS will create an internal DNS record for `my-pod.internal.example.com` targeting the Pod `Status.PodIP`.

## external-dns.alpha.kubernetes.io/resolve-target

> Controls whether load balancer hostname targets when they are CNAME records. When the annotation is set to `"true"` the CNAME records are resolved to their underlying

Reviewer: ```suggestion
Controls load balancer hostname targets with CNAME records. 
When the annotation is set to `"true"` the CNAME records are resolved to their underlying
```

A/AAAA records at sync time. This is useful when a DNS provider does not support CNAME flattening
or ALIAS records and a flat IP address record is required.

Supported values:

- `"true"` — resolve the resource's hostname targets to A/AAAA records.
- `"false"` — keep the resource's hostname targets as CNAME records.

If the annotation is absent, hostname targets are kept as CNAME records (the default behaviour).

When resolution is enabled, ExternalDNS performs a DNS lookup for each hostname target and emits the
resulting IP addresses as A and/or AAAA endpoints. If resolution fails (e.g. the hostname is
temporarily unresolvable), that target is silently skipped.

### Use Cases for `external-dns.alpha.kubernetes.io/resolve-target` annotation

#### Opt in to resolution for a single Source

Resolve load balancer hostnames to IPs for one Ingress:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: my-ingress
  annotations:

>     external-dns.alpha.kubernetes.io/resolve-target: "true"

Author: fixed in d8223dc843d601225a26dfa4a8c771d454e6c3bf
